process/block/preprocess/transactions.go

```diff
@@ -472,151 +718,151 @@ func (txs *transactions) getAllTxsFromMiniBlock(
 
 // CreateAndProcessMiniBlocks creates miniblocks from storage and processes the transactions added into the miniblocks
 // as long as it has time
-func (txs *transactions) CreateAndProcessMiniBlocks(
-	maxTxSpaceRemained uint32,
-	maxMbSpaceRemained uint32,
-	haveTime func() bool,
-) (block.MiniBlockSlice, error) {
-
-	miniBlocks := make(block.MiniBlockSlice, 0)
-	newMBAdded := true
-	txSpaceRemained := int(maxTxSpaceRemained)
-
-	miniBlock, err := txs.createAndProcessMiniBlock(
-		txs.shardCoordinator.SelfId(),
-		core.MetachainShardId,
-		txSpaceRemained,
-		haveTime)
-
-	if err == nil && len(miniBlock.TxHashes) > 0 {
-		txSpaceRemained -= len(miniBlock.TxHashes)
-		miniBlocks = append(miniBlocks, miniBlock)
+func (txs *transactions) CreateAndProcessMiniBlocks(haveTime func() bool) (block.MiniBlockSlice, error) {
+	timeBefore := time.Now()
+	sortedTxs, err := txs.computeSortedTxs(txs.shardCoordinator.SelfId(), txs.shardCoordinator.SelfId())
+	timeAfter := time.Now()
+	if err != nil {
+		log.Debug("computeSortedTxs", "error", err.Error())
+		return make(block.MiniBlockSlice, 0), nil
 	}
 
-	for newMBAdded {
-		newMBAdded = false
-		for shardId := uint32(0); shardId < txs.shardCoordinator.NumberOfShards(); shardId++ {
-			if !haveTime() {
-				break
-			}
-
-			mbSpaceRemained := int(maxMbSpaceRemained) - len(miniBlocks)
-			if mbSpaceRemained <= 0 {
-				break
-			}
+	if !haveTime() {
+		log.Debug("time is up ordering txs",
+			"num txs", len(sortedTxs),
+			"time [s]", timeAfter.Sub(timeBefore).Seconds(),
+		)
+		return make(block.MiniBlockSlice, 0), nil
+	}
 
-			//TODO: We should analyze if this check could be done more restrictive or not, depending of the pending
-			//miniblocks given by the last metablock notarized instead of pending miniblocks given by the last metablock
-			//received in block tracker (the current state of all shards)
-			if txs.blockTracker.IsShardStuck(shardId) {
-				continue
-			}
+	log.Debug("time elapsed to ordered txs",
+		"num txs", len(sortedTxs),
+		"time [s]", timeAfter.Sub(timeBefore).Seconds(),
+	)
 
-			var miniBlockForShard *block.MiniBlock
-			miniBlockForShard, err = txs.createAndProcessMiniBlock(
-				txs.shardCoordinator.SelfId(),
-				shardId,
-				txSpaceRemained,
-				haveTime)
-			if err != nil {
-				continue
-			}
+	startTime := time.Now()
+	miniBlocks, err := txs.createAndProcessMiniBlocks(haveTime, txs.blockTracker.IsShardStuck, sortedTxs)
```

Comment:
no need to pass the block tracker is shard stuck function. it can be called inside the txs,createAndProcessMiniBlocks function directly. txs.blockTracer.IsShardStuck(shId)

Reply:
As this method is used now also in the validator side, just to have the same pattern as the proposer, we should avoid to check if shard is stuck when node is validator as the view could be different depending of the time when this is checked (more problems when nodes are syncing)

Reply:
ok. agreed.